### PR TITLE
#3586-UI-CV- add-edit_rh_custom_spin

### PR DESCRIPTION
**Pull request by ldjebran**
``` bash
dlezz@elysion:~/projects/robottelo-fork$ py.test tests/foreman/ui/test_contentview.py -v -k 'test_positive_edit_rh_custom_spin'
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.12, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- /home/dlezz/bin/redhat/python/2.7/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile: 
plugins: cov-2.4.0, xdist-1.15.0
collected 65 items 

tests/foreman/ui/test_contentview.py::ContentViewTestCase::test_positive_edit_rh_custom_spin <- robottelo/decorators/__init__.py PASSED

================================================= 64 tests deselected ==================================================
====================================== 1 passed, 64 deselected in 306.83 seconds =======================================
```

``` bash
dlezz@elysion:~/projects/robottelo-fork$ py.test tests/foreman/ui/test_contentview.py -v -k 'test_positive_add_rh_custom_spin'
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.12, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- /home/dlezz/bin/redhat/python/2.7/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile: 
plugins: cov-2.4.0, xdist-1.15.0
collected 65 items 

tests/foreman/ui/test_contentview.py::ContentViewTestCase::test_positive_add_rh_custom_spin <- robottelo/decorators/__init__.py PASSED

================================================= 64 tests deselected ==================================================
====================================== 1 passed, 64 deselected in 304.04 seconds =======================================
```

### CODE REVIEW

File: robottelo/ui/contentviews.py [hunk 1]
```diff
@@ -30,6 +30,18 @@ def go_to_filter_page(self, cv_name, filter_name):
         strategy, value = locators['contentviews.select_filter_name']
         self.click((strategy, value % filter_name))
 
+    def set_calendar_date_value(self, name, value):
+        """Set the input value of a date field and press the button to hide
+         the calendar popup panel"""
+        date_elm = self.wait_until_element(
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: `date_element` is definitely better naming

File: robottelo/ui/contentviews.py [hunk 2]
```diff
@@ -30,6 +30,18 @@ def go_to_filter_page(self, cv_name, filter_name):
...
+            locators['contentviews.calendar_date_input'] % name)
+        date_elm.clear()
+        if value:
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: That doesn't make sense as value cannot be None according to method signature

Comment (oshtaier) [inline robottelo/ui/contentviews.py]: That is why we don't need all these statements and can use simple:
`text_field_update` method from base class

File: robottelo/ui/contentviews.py [hunk 3]
```diff
@@ -450,6 +462,41 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
         else:
             self.click(locators['contentviews.remove_errata'])
 
+    def edit_erratum_date_range_filter(
+            self, cv_name, filter_name, errata_types=None, date_type=None,
+            start_date=None, end_date=None, open_filter=True):
+        """Edit Erratum Date Range Filter"""
+        allowed_errata_types = ['security', 'enhancement', 'bugfix']
+        allowed_date_types = ['updated', 'issued']
+        if not errata_types:
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: hmm, it seems that `errata_types=['security']` in method signature will do the same trick

File: robottelo/ui/contentviews.py [hunk 4]
```diff
@@ -450,6 +462,41 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+            errata_types = ['security']
+        if not date_type:
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: same here

File: robottelo/ui/contentviews.py [hunk 5]
```diff
@@ -450,6 +462,41 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+            self.go_to_filter_page(cv_name, filter_name)
+        for errata_type in allowed_errata_types:
+            errata_type_elm = self.wait_until_element(
+                locators['contentviews.erratum_type_checkbox'] % errata_type)
+            if errata_type_elm:
+                if errata_type_elm.get_attribute('value') == 'true':
+                    checked = True
+                else:
+                    checked = False
+                if (errata_type in errata_types and not checked)\
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: I would say that `\`  style is not recommended at all. Please try:
```
if ((errata_type in errata_types and not checked) or 
                errata_type  not in errata_types and checked)):
```

File: robottelo/ui/contentviews.py [hunk 6]
```diff
@@ -450,6 +462,41 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+        if not errata_types:
+            errata_types = ['security']
+        if not date_type:
+            date_type = 'updated'
+        if date_type not in allowed_date_types:
+            raise UIError('date type "{0}" does not exit'.format(date_type))
+        if open_filter:
+            self.go_to_filter_page(cv_name, filter_name)
+        for errata_type in allowed_errata_types:
+            errata_type_elm = self.wait_until_element(
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: `errata_type_element` seems as better naming

File: robottelo/ui/contentviews.py [hunk 7]
```diff
@@ -450,6 +462,41 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+                locators['contentviews.erratum_type_checkbox'] % errata_type)
+            if errata_type_elm:
+                if errata_type_elm.get_attribute('value') == 'true':
```
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: Can we use `self.wait_until_element(locator).is_selected()` instead here? in case that is a checkbox of course. Then probably we will not need that if condition

File: tests/foreman/ui/test_contentview.py [hunk 8]
```diff
@@ -464,11 +467,49 @@ def test_positive_edit_rh_custom_spin(self):
...
+            self.assertIsNotNone(
+                self.content_views.search_filter(cv_name, filter_name))
+            # go to the filter open it, edit it and save
+            self.content_views.edit_erratum_date_range_filter(
+                cv_name,
+                filter_name,
+                errata_types=[FILTER_ERRATA_TYPE['enhancement'],
+                              FILTER_ERRATA_TYPE['bugfix']],
+                date_type=FILTER_ERRATA_DATE['issued'],
+                start_date='2016-01-01',
```
Comment (oshtaier) [inline tests/foreman/ui/test_contentview.py]: are these dates gonna be valid in future? In any moment?

File: tests/foreman/ui/test_contentview.py [hunk 9]
```diff
@@ -611,11 +652,53 @@ def test_positive_add_rh_custom_spin(self):
...
+            self.content_views.add_remove_repos(cv_name, [rh_repo['name']])
+            self.assertIsNotNone(self.content_views.wait_until_element(
+                common_locators['alert.success_sub_form']))
+            self.content_views.add_filter(
+                cv_name,
+                filter_name,
+                FILTER_CONTENT_TYPE['erratum by date and type'],
+                FILTER_TYPE['exclude']
+            )
+            # edit the filter fields immediately after creation and without
```
Comment (oshtaier) [inline tests/foreman/ui/test_contentview.py]: Can you rephrase this comment as I didn't get it?

**Comment by oshtaier** [comment]
Pending comments

File: robottelo/ui/contentviews.py [hunk 10]
```diff
@@ -30,6 +30,21 @@ def go_to_filter_page(self, cv_name, filter_name):
...
         self.click((strategy, value % filter_name))
 
+    def set_calendar_date_value(self, name, value=None):
+        """Set the input value of a date field and press the button to hide
+         the calendar popup panel"""
+        date_element = self.wait_until_element(
+            locators['contentviews.calendar_date_input'] % name)
+        date_element.clear()
+        if value:
+            date_element.send_keys(value)
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: So if value was passed, we're setting it, if not (meaning user don't want to touch it) - we're clearing existing value and overwriting it with empty value, which is probably not something user will be happy about.
I'd rather
a) clear `date_element` only in case we're going to change it
b) changed `if value:` to `if value is not None:`. This way if user wants to clear the value, he'll be able to simply pass empty string `'' ` as value.

File: tests/foreman/ui/test_contentview.py [hunk 11]
```diff
@@ -611,11 +652,53 @@ def test_positive_add_rh_custom_spin(self):
...
+            # will edit the filter fields immediately with open_filter=False
+            self.content_views.edit_erratum_date_range_filter(
+                cv_name,
+                filter_name,
+                errata_types=[FILTER_ERRATA_TYPE['enhancement'],
+                              FILTER_ERRATA_TYPE['bugfix']],
+                date_type=FILTER_ERRATA_DATE['issued'],
+                start_date='2016-01-01',
+                end_date='2016-06-01',
+                open_filter=False
```
Comment (abalakh) [inline tests/foreman/ui/test_contentview.py]: I think in 99% cases our helpers are isolated and each is not relying on previous steps but opening entity by its own.  Not sure if there's a need in option to open entity or rely on user. @oshtaier , your thoughts?

File: tests/foreman/ui/test_contentview.py [hunk 12]
```diff
@@ -611,11 +652,53 @@ def test_positive_add_rh_custom_spin(self):
...
+                common_locators['alert.success_sub_form']))
+            self.content_views.add_filter(
+                cv_name,
+                filter_name,
+                FILTER_CONTENT_TYPE['erratum by date and type'],
+                FILTER_TYPE['exclude']
+            )
+            # The last executed function will create the filter and open the
+            # edit filter form.
+            # will not assert the filter existence here to not exist from the
```
Comment (ldjebran) [inline tests/foreman/ui/test_contentview.py]: this is a typo:  exit (to be updated)

File: robottelo/ui/contentviews.py [hunk 13]
```diff
@@ -450,6 +465,37 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
         else:
             self.click(locators['contentviews.remove_errata'])
 
+    def edit_erratum_date_range_filter(
+            self, cv_name, filter_name, errata_types=None, date_type='updated',
+            start_date=None, end_date=None, open_filter=True):
+        """Edit Erratum Date Range Filter"""
+        allowed_errata_types = ['security', 'enhancement', 'bugfix']
```
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: you are right, do not wanted to make an extra import (to update)

File: robottelo/ui/contentviews.py (hunk 10 above)
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: this is a set value function, if a user do want to change the value it have to not set it's value (by calling this function), will redo back this signature for value without default value None

Comment (oshtaier) [inline robottelo/ui/contentviews.py]: Yeah, yesterday, I understood that we always wanted to clear that value, but if no and that is usual input, so we need to remove entire code and use `text_field_update` as I suggested from the start. And if user wanted to clear, he will send `''` and it will be cleared

File: robottelo/ui/contentviews.py [hunk 14]
```diff
@@ -450,6 +465,37 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+        allowed_date_types = ['updated', 'issued']
+        if not errata_types:
+            # one errata type is mandatory
+            errata_types = ['security']
```
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: this function will be changed for None value to be not updated,
but still if errata_types is not None it has to have one element as one element in the list is mandatory (open the filter and you will realize the UI will not allow all the check boxes to be unchecked)

File: robottelo/ui/contentviews.py [hunk 15]
```diff
@@ -450,6 +465,37 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+                if ((errata_type in errata_types and not checked) or
+                        (errata_type not in errata_types and checked)):
+                    # click on it only if it's checked and not in the selection
+                    # or not checked and in the selection
+                    errata_type_element.click()
+            else:
+                raise UIError(
+                    'Cannot find the check box element %s .', errata_type)
+        self.click(locators['contentviews.erratum_date_type'] % date_type)
+        self.set_calendar_date_value('start_date', start_date)
```
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: set_calendar_date_value will be changed (see above comment) 
and this function will be changed to not update the value with None values

File: tests/foreman/ui/test_contentview.py (hunk 11 above)
Comment (ldjebran) [inline tests/foreman/ui/test_contentview.py]: in this special case: we have two test functions  add and edit, we have 2 solutions.
1 - Follow the flow of adding a filter (by not reopening the filter form), and here to not write the same code that differ only by one line self.go_to_filter_page(cv_name, filter_name), I added this parameter, open_filter
2 - Do not follow the flow of adding a filter, this way , we have to delete this test function as it will the same as edit test function.

**Comment by ldjebran** [comment]
@oshtaier , @abalakh thank you for comments, refactored the code especialy the functions in base contentviews, thanks to review

File: robottelo/ui/contentviews.py [hunk 16]
```diff
@@ -450,6 +462,45 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+                # because of the behaviour of the UI to check and disable
+                # the last element if the others are unchecked,
+                # will check all the elements first,
+                # after then uncheck the not selected ones
+                for errata_type in allowed_errata_types:
+                    self.assign_value(erratum_type_locator % errata_type, True)
+                for errata_type in allowed_errata_types:
+                    if errata_type not in errata_types:
+                        self.assign_value(
+                            erratum_type_locator % errata_type, False)
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: NACK, this is basically wrong. 2 loops for just checking checkboxes is not ok.
Think of this:
```python
for errata in allowed_errata_types:
    self.assign_value(
        locators['contentviews.erratum_type_checkbox'] % checkbox,
        errata in errata_types
    )
```
So instead of firstly checking everything on and then unchecking what's not supposed to be checked you will set correct state for each checkbox one by one

File: robottelo/ui/contentviews.py [hunk 17]
```diff
@@ -450,6 +462,45 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+            else:
+                # minimum one errata type is mandatory
+                raise UIError(
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: Current style:
```python
if errata_types:
    # line of code
    # another one
    # and one more
else:
    raise IUError()
```
adds redundant indentation for all the actual logic of your method. Prefer following one instead:
```python
if not errata_types:
    raise UIError()
# line of code
# another one
# and one more

File: robottelo/ui/contentviews.py [hunk 18]
```diff
@@ -450,6 +462,45 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+                # the last element if the others are unchecked,
+                # will check all the elements first,
+                # after then uncheck the not selected ones
+                for errata_type in allowed_errata_types:
+                    self.assign_value(erratum_type_locator % errata_type, True)
+                for errata_type in allowed_errata_types:
+                    if errata_type not in errata_types:
+                        self.assign_value(
+                            erratum_type_locator % errata_type, False)
+            else:
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: OK. So i've passed `errata_types=['foo']`. This way, it's not `None`, it's not empty (i.e. `True` for second check), but nothing will be actually checked off and exception won't be risen. Think of different check instead of just `if errata_types:`

Comment (abalakh) [inline robottelo/ui/contentviews.py]: tip: I'd rather use sets, `.issubset()` to be more specific. It will compare inputs for us and in case something is wrong in there, or if input list is empty - you can raise an exception

File: robottelo/ui/contentviews.py [hunk 19]
```diff
@@ -450,6 +462,45 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
...
+                # minimum one errata type is mandatory
+                raise UIError(
+                    'errata types is empty, minimum required: one errata type')
+        if date_type is not None:
+            if date_type in allowed_date_types:
+                self.click(date_type_locator % date_type)
+            else:
+                raise UIError(
+                    'date type "{0}" not allowed'.format(date_type))
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: Same comment about style

File: robottelo/ui/contentviews.py (hunk 16 above)
Comment (abalakh) [inline robottelo/ui/contentviews.py]: Ok, so after discussion it appears that the last checked off checkbox can't be unchecked - it becomes disabled till you check off something else. So if we had `bugfix` errata and wan't to update it to `enhancement` - we shall check off enhancement first and only then uncheck bugfix. That explains 2 loops here.
With that said, i can see 2 solutions here:
1) 2 loops, one checking off elements, second one is unchecking. This is basically what's applied here, my only suggestion is in the first loop not to check off everything, but only those which need to be checked.
2) Form a list of checkboxes in the way that the ones which need to be checked will appear first. This way we can use only 1 loop:
```python
for errata in errata_types + list(set(allowed_errata_types).difference(errata_types)):
    self.assign_value(
        locators['contentviews.erratum_type_checkbox'] % checkbox,
        errata in errata_types
    )
```
The only problem with this approach - it looks a bit overcomplicated. Or it's just me? Any other reviewer's thoughts?

**Comment by ldjebran** [comment]
@oshtaier , @abalakh, any comments for last version, thanks in advance

File: robottelo/ui/contentviews.py [hunk 20]
```diff
@@ -35,12 +35,12 @@ def set_calendar_date_value(self, name, value):
         """Set the input value of a date field and press the button to hide
          the calendar popup panel"""
         self.assign_value(
-            locators['contentviews.calendar_date_input'] % name, value)
+            locators.contentviews.calendar_date_input % name, value
+        )
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: nitpick:
either
```python
self.assign_value(
    locators.contentviews.calendar_date_input % name, value)
```
or
```python
self.assign_value(
    locators.contentviews.calendar_date_input % name,
    value
)
```
not a blocker

File: robottelo/ui/contentviews.py [hunk 21]
```diff
@@ -450,6 +462,50 @@ def add_remove_errata_to_filter(self, cv_name, filter_name,
         else:
             self.click(locators['contentviews.remove_errata'])
 
+    def edit_erratum_date_range_filter(
+            self, cv_name, filter_name, errata_types=None, date_type=None,
+            start_date=None, end_date=None, open_filter=True):
+        """Edit Erratum Date Range Filter"""
+        allowed_errata_types = FILTER_ERRATA_TYPE.values()
+        allowed_date_types = FILTER_ERRATA_DATE.values()
```
Comment (abalakh) [inline robottelo/ui/contentviews.py]: after all the comments applied `allowed_date_types` became single-use variable, presence of which we're trying to avoid.
not a blocker too, and basically it's debatable as this specific single-use variable adds readability.

File: robottelo/ui/contentviews.py (hunk 20 above)
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: I looked in the code in this file, any where were two variables it is like:  
```python
self.assign_value(
    locators.contentviews.calendar_date_input % name,
    value
)
```

**Comment by abalakh** [comment]
ACK

File: robottelo/ui/contentviews.py (hunk 20 above)
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: ok, understood what you mean thanks

File: robottelo/ui/contentviews.py (hunk 21 above)
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: + for readability

File: robottelo/ui/contentviews.py (hunk 20 above)
Comment (ldjebran) [inline robottelo/ui/contentviews.py]: resolved

**Comment by ldjebran** [comment]
@oshtaier all the necessary changes has been done, thanks to review.

File: tests/foreman/ui/test_contentview.py (hunk 11 above)
Comment (oshtaier) [inline tests/foreman/ui/test_contentview.py]: So... Do edit and add test methods has only one difference in whether we opening filter page or not? Not sure that I understand that, because editing and adding functionalities seem as pretty different things in logical way

**Comment by sghai** [comment]
@ldjebran : Hey does this test asserts the excluded contents ? I think that's what we need to validate.

@oshtaier: assuming all your comments have been addressed ?

**Comment by ldjebran** [comment]
@sghai, No this test is about the content view filters and by design in docs and by corresponding tests in api does not need to assert that the content has been excluded, the excluded type was my choice.

**Comment by ldjebran** [comment]
@sghai, these tests are about the ability to add and edit content view filters

**Comment by ldjebran** [comment]
@sghai, the problem was not in the test it self, but in the function in the base function that had to be more general, and that has been fixed with Andrei  review

**Comment by sghai** [comment]
@ldjebran : Alright. thanks for the update. Merging now.
